### PR TITLE
Add support to expose/redirect/disable insecure schemes (http) for edge secured routes.

### PR DESCRIPTION
**Pull request by ramr**
This allows secure and non-secure routes for the same host. Fix for #4848 

@smarterclayton / @pweil-   PTAL  Thx

New routes can contain an insecure field w/ supported values `Redirect | Allow | None` - default is None (or empty string for older routes). 
Example route to redirect http to https:
```
{  
  "kind": "Route",  
  "apiVersion": "v1beta3",  
  "metadata": {  
    "name": "header-test-edge-redirect-route"  
  },  
  "spec": {  
    "host": "edge.header.test",  
    "tls": {  
      "termination": "edge",  
      "insecureEdgeTerminationPolicy": "Redirect"  
    },  
    "to": {  
      "name": "header-test"  
    }  
  }  
}  
``` 

```
[vagrant@openshiftdev origin]$ oc create -f test-route-redirect.json  
route "header-test-edge-redirect-route" created  
[vagrant@openshiftdev origin]$ curl -vvv -k --resolve edge.header.test:80:127.0.0.1 http://edge.header.test  
... <snipped> ...   
< HTTP/1.1 302 Found  
< Cache-Control: no-cache  
< Content-length: 0  
< Location: https://edge.header.test/  
< Connection: close  
<   
* Closing connection 0  
```
 
*Updated:  Updated examples to use the new constants / field names.

### CODE REVIEW

**Comment by mfojtik** [comment]
@bparees FYI, I hope this won't break the Jenkins again :-)

**Comment by liggitt** [comment]
special casing insecure as a redirect to secure seems like a really narrow special case

**Comment by smarterclayton** [comment]
Agree.

File: pkg/route/api/types.go [hunk 1]
```diff
@@ -89,11 +89,19 @@ type TLSConfig struct {
 	// DestinationCACertificate provides the contents of the ca certificate of the final destination.  When using reencrypt
 	// termination this file should be provided in order to have routers use it for health checks on the secure connection
 	DestinationCACertificate string `json:"destinationCACertificate,omitempty"`
+
+	// Insecure indicates the desired behavior for insecure connections
+	// to an edge-terminated route: expose, redirect or disable.
+	Insecure TLSInsecureType `json:"insecure,omitempty"`
```
Comment (smarterclayton) [inline pkg/route/api/types.go]: `InsecurePolicy`

File: pkg/route/api/types.go [hunk 2]
```diff
@@ -101,4 +109,11 @@ const (
 	TLSTerminationPassthrough TLSTerminationType = "passthrough"
 	// TLSTerminationReencrypt terminate encryption at the edge router and re-encrypt it with a new certificate supplied by the destination
 	TLSTerminationReencrypt TLSTerminationType = "reencrypt"
+
+	// TLSInsecureDisable disables insecure connections for an edge-terminated route.
+	TLSInsecureDisable TLSInsecureType = "disable"
```
Comment (smarterclayton) [inline pkg/route/api/types.go]: All constants should be `CamelCase` - don't change the old ones, but any new ones should follow that rule.

File: pkg/route/api/validation/validation.go [hunk 3]
```diff
@@ -135,3 +140,35 @@ func validateNoDoubleEscapes(tls *routeapi.TLSConfig) fielderrors.ValidationErro
 	}
 	return allErrs
 }
+
+// validateTLSInsecure tests fields for different types of insecure options.
+// Called by validateTLS.
+func validateTLSInsecure(tls *routeapi.TLSConfig) *fielderrors.ValidationError {
+	// Check insecure option value if specified (empty is ok).
+	if tls == nil || tls.Insecure == "" {
```
Comment (smarterclayton) [inline pkg/route/api/validation/validation.go]: `len(tls.Insecure) == 0`

File: pkg/route/api/v1/types.go [hunk 4]
```diff
@@ -106,4 +114,11 @@ const (
 	TLSTerminationPassthrough TLSTerminationType = "passthrough"
 	// TLSTerminationReencrypt terminate encryption at the edge router and re-encrypt it with a new certificate supplied by the destination
 	TLSTerminationReencrypt TLSTerminationType = "reencrypt"
+
+	// TLSInsecureDisable disables insecure connections for an edge-terminated route.
+	TLSInsecureDisable TLSInsecureType = "disable"
```
Comment (smarterclayton) [inline pkg/route/api/v1/types.go]: `None`

File: pkg/route/api/v1/types.go [hunk 5]
```diff
@@ -106,4 +114,11 @@ const (
...
+	// TLSInsecureExpose allows insecure connections for an edge-terminated route.
+	TLSInsecureExpose TLSInsecureType = "expose"
```
Comment (smarterclayton) [inline pkg/route/api/v1/types.go]: `Allow`

File: pkg/route/api/v1/types.go [hunk 6]
```diff
@@ -106,4 +114,11 @@ const (
...
+	// TLSInsecureRedirect redirects insecure connections for an edge-terminated route.
```
Comment (smarterclayton) [inline pkg/route/api/v1/types.go]: Expand the description to give an example - for instance, "Routers that support HTTP will redirect the connection to HTTPS"

File: pkg/route/api/validation/validation.go [hunk 7]
```diff
@@ -135,3 +140,35 @@ func validateNoDoubleEscapes(tls *routeapi.TLSConfig) fielderrors.ValidationErro
...
+		return nil
+	}
+
+	// Ensure insecure is set only for edge terminated routes.
+	if routeapi.TLSTerminationEdge != tls.Termination {
+		// tls.Insecure option is not supported for a non edge-terminated routes.
+		msg := fmt.Sprintf("invalid value %s for insecure option - insecure is only supported for edge-terminated routes", tls.Insecure)
```
Comment (liggitt) [inline pkg/route/api/validation/validation.go]: way too verbose, the current value is already included... `fielderrors.NewFieldInvalid("insecure", tls.Insecure, "insecure is only allowed for edge-terminated routes")`

File: pkg/route/api/validation/validation.go [hunk 8]
```diff
@@ -135,3 +140,35 @@ func validateNoDoubleEscapes(tls *routeapi.TLSConfig) fielderrors.ValidationErro
...
+		return fielderrors.NewFieldInvalid("insecure", tls.Insecure, msg)
+	}
+
+	// It is an edge-terminated route, check insecure option value is
+	// one of Disable/Expose/Redirect.
+	switch tls.Insecure {
```
Comment (liggitt) [inline pkg/route/api/validation/validation.go]: put the supported values in a string set, check if it's in that set, then do NewFieldValueNotSupported with the allowed values

**Comment by ramr** [comment]
@liggitt / @smarterclayton  will address your comments when I get some time this week. Thx

**Comment by smarterclayton** [comment]
We may need to make it possible to direct insecure to its own port.  I'm
torn on this but I don't like separating routes by port until ingress lands.

On Oct 20, 2015, at 1:49 PM, Ram Ranganathan <notifications@github.com>
wrote:

@liggitt <https://github.com/liggitt> / @smarterclayton
<https://github.com/smarterclayton> will address your comments when I get
some time this week. Thx

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5224#issuecomment-149645742>.

**Comment by smarterclayton** [comment]
Or we use tls state +insecure to set the rules.  It's pretty ugly for sure.

On Oct 20, 2015, at 1:49 PM, Ram Ranganathan <notifications@github.com>
wrote:

@liggitt <https://github.com/liggitt> / @smarterclayton
<https://github.com/smarterclayton> will address your comments when I get
some time this week. Thx

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5224#issuecomment-149645742>.

File: images/router/haproxy-base/Dockerfile [hunk 9]
```diff
@@ -14,7 +14,7 @@ FROM openshift/origin-base
 RUN yum -y install haproxy && \
     mkdir -p /var/lib/containers/router/{certs,cacerts} && \
     mkdir -p /var/lib/haproxy/{conf,run,bin,log} && \
-    touch /var/lib/haproxy/conf/{{os_http_be,os_edge_http_be,os_tcp_be,os_sni_passthrough,os_reencrypt}.map,haproxy.config} && \
+    touch /var/lib/haproxy/conf/{{os_http_be,os_edge_http_be,os_tcp_be,os_sni_passthrough,os_reencrypt,os_edge_http_expose,os_edge_http_redirect}.map,haproxy.config} && \
```
Comment (smarterclayton) [inline images/router/haproxy-base/Dockerfile]: Do we still have to do this?

**Comment by mrry550** [comment]
@ramr @smarterclayton - do we need to do this for 3.1?

**Comment by smarterclayton** [comment]
I would like to say no due to complexity.  Given the time window I don't
see the risk being high enough.  We should bring it up at the next triage
meeting - in the meantime Ram if you can update this if it's not taking
higher priority than your other items.

**Comment by mrry550** [comment]
@smarterclayton - thanks! Appreciate it. :)

**Comment by ramr** [comment]
@smarterclayton am actually OOO till Thu but will take care of it tomorrow am.

**Comment by smarterclayton** [comment]
Don't worry about it.

File: pkg/route/api/types.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/route/api/types.go]: `InsecureEdgeTerminationPolicy` and `insecureEdgeTerminationPolicy` - I'd rather be a bit more explanatory

**Comment by smarterclayton** [comment]
Can you address the comments today?

**Comment by smarterclayton** [comment]
Add a unit test after you address the commits

**Comment by ramr** [comment]
Yeah unfortunately was debugging some issues QE was facing all day. I will try and address this tonight but it might slip to sometime tomorrow.  Thx

File: images/router/haproxy-base/Dockerfile (hunk 9 above)
Comment (ramr) [inline images/router/haproxy-base/Dockerfile]: Not really but it does tell you all the maps there are. I think we wanted  to get rid of this image and so will end up having to merge into the haproxy dockerfile - fold it into  ```images/router/haproxy/Dockerfile```.

**Comment by ramr** [comment]
@smarterclayton  - made the changes as per review comments above + added another unit test in the template router (there's one in validation and one in f5 that were already there).

**Comment by ramr** [comment]
@smarterclayton  - fyi did the requested changes yesterday + build passed.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**